hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java

```diff
@@ -1373,4 +1376,109 @@ protected void validateEncryptionKeyInfo(OmBucketInfo bucketInfo, KeyArgs keyArg
+    checkBucketQuotaInNamespace(omBucketInfo, missingParentInfos.size());
+    omBucketInfo.incrUsedNamespace(missingParentInfos.size());
+
+    // Add cache entries for the missing parent directories.
+    OMFileRequest.addDirectoryTableCacheEntries(omMetadataManager,
+            volumeId, bucketId, transactionLogIndex,
+            missingParentInfos, null);
+  }
+
+  protected OmKeyInfo getOmKeyInfoFromOpenKeyTable(String dbMultipartKey,
+                                                   String keyName,
+                                                   OMMetadataManager omMetadataManager) throws IOException {
+    return omMetadataManager.getOpenKeyTable(getBucketLayout())
+            .get(dbMultipartKey);
+  }
+
+  protected void addMultiPartToCache(
+          OMMetadataManager omMetadataManager, String multipartOpenKey,
+          OMFileRequest.OMPathInfoWithFSO pathInfoFSO, OmKeyInfo omKeyInfo,
+          String keyName, long transactionLogIndex
+  ) {
+
+    // Add multi part to cache
+    OMFileRequest.addOpenFileTableCacheEntry(omMetadataManager,
+            multipartOpenKey, omKeyInfo, pathInfoFSO.getLeafNodeName(),
+            keyName, transactionLogIndex);
+
+  }
+
+  protected List<OmDirectoryInfo> addMissingDirectories(OzoneManager ozoneManager,
+                                                        OzoneManagerProtocolProtos.KeyArgs keyArgs,
+                                                        long trxnLogIndex) throws
+          IOException {
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    final String volumeName = keyArgs.getVolumeName();
+    final String bucketName = keyArgs.getBucketName();
+    final String keyName = keyArgs.getKeyName();
+    OmBucketInfo omBucketInfo = getBucketInfo(omMetadataManager,
+            volumeName, bucketName);
+    List<OmDirectoryInfo> missingParentInfos;
+    OMFileRequest.OMPathInfoWithFSO pathInfoFSO = OMFileRequest
+            .verifyDirectoryKeysInPath(omMetadataManager, volumeName, bucketName,
+                    keyName, Paths.get(keyName));
+    missingParentInfos = getAllMissingParentDirInfo(ozoneManager, keyArgs, omBucketInfo,
+                    pathInfoFSO, trxnLogIndex);
+
+    if (missingParentInfos != null && !missingParentInfos.isEmpty()) {
+      final long volumeId = omMetadataManager.getVolumeId(volumeName);
+      final long bucketId = omMetadataManager.getBucketId(volumeName,
+              bucketName);
+
+      // add all missing parents to directory table
+      addMissingParentsToCache(omBucketInfo, missingParentInfos,
+              omMetadataManager, volumeId, bucketId, trxnLogIndex);
+
+      String multipartOpenKey = omMetadataManager
+              .getMultipartKey(volumeId, bucketId,
+                      pathInfoFSO.getLastKnownParentId(),
+                      pathInfoFSO.getLeafNodeName(),
+                      keyArgs.getMultipartUploadID());
```

Comment:
nit: Please indent continuations with 4 spaces, not 8.  (Applies to the patch as a whole, this is only an example.)

Reply:
reformatted